lib/OpenQA/WebAPI/Plugin/Fedmsg.pm

```diff
@@ -0,0 +1,76 @@
+# Copyright (C) 2016 Red Hat
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, see <http://www.gnu.org/licenses/>.
+
+package OpenQA::WebAPI::Plugin::Fedmsg;
+
+use strict;
+use warnings;
+
+use parent qw/Mojolicious::Plugin/;
+use IPC::Run;
+use JSON;
+use Mojo::IOLoop;
+use OpenQA::Schema::Result::Jobs;
+
+my @events = qw/job_create job_delete job_cancel job_duplicate job_restart jobs_restart job_update_result job_done/;
+
+sub register {
+    my ($self, $app) = @_;
+    my $reactor = Mojo::IOLoop->singleton;
+
+    # register for events
+    for my $e (@events) {
+        $reactor->on("openqa_$e" => sub { shift; $self->on_event($app, @_) });
+    }
+}
+
+# when we get an event, convert it to fedmsg format and send it
+sub on_event {
+    my ($self, $app, $args) = @_;
+    my ($user_id, $connection_id, $event, $event_data) = @$args;
+    # we're going to explicitly pass this as the modname
+    $event =~ s/^openqa_//;
+    # fedmsg uses dot separators
+    $event =~ s/_/\./;
+    # find count of pending jobs for the same build
+    # this is so we can tell when all tests for a build are done
+    my $build = $app->db->resultset('Jobs')->find({id => $event_data->{id}})->settings_hash->{BUILD};
+    $event_data->{remaining} = $app->db->resultset('Jobs')->search(
+        {
+            'settings.key'   => 'BUILD',
+            'settings.value' => $build,
+            state            => [OpenQA::Schema::Result::Jobs::PENDING_STATES],
+        },
+        {join => qw/settings/})->count;
+    # add a 'build' property to the event data, this is handy for consumers
+    $event_data->{build} = $build;
+    # there seems to be some weird fuzziness about whether job IDs show
+    # up as integers or strings in the JSON which messes with the tests,
+    # so lets try and force them to integers
+    $event_data->{id}     = $event_data->{id} + 0     if ($event_data->{id}     && $event_data->{id} =~ m/\d+/);
+    $event_data->{result} = $event_data->{result} + 0 if ($event_data->{result} && $event_data->{result} =~ m/\d+/);
```

Comment:
The perl JSON reference may help here - http://search.cpan.org/~makamaka/JSON-2.90/lib/JSON.pm#PERL_-%3E_JSON . It seems like if the scalar has been used in a string context it winds up as a string in the JSON. But you probably know more about perl's handling of this stuff than I do...if you can fix it, great.